libcanard/canard.c

```diff
@@ -547,6 +592,35 @@ CANARD_PRIVATE int32_t txPushMultiFrame(struct CanardTxQueue* const        que,
     return out;
 }
 
+/// Flushes expired transfers by comparing deadline timestamps of the pending transfers with the current time.
+CANARD_PRIVATE void txFlushExpiredTransfers(struct CanardTxQueue* const        que,
+                                            const struct CanardInstance* const ins,
+                                            const CanardMicrosecond            now_usec)
+{
+    struct CanardTxQueueItem* tx_item = NULL;
+    while (NULL != (tx_item = MUTABLE_CONTAINER_OF(  //
+                        struct CanardTxQueueItem,
+                        cavlFindExtremum(que->deadline_root, false),
+                        deadline_base)))
+    {
+        if (now_usec <= tx_item->tx_deadline_usec)
+        {
+            // The queue is sorted by deadline, so we can stop here.
+            break;
+        }
+
+        // All frames of the transfer are released at once b/c they all have the same deadline.
```

Comment:
Do we really need this special case here? All frames in a transfer have the same timeout (IIRC it is mentioned in the docs, too), so either all or none will be kept even if you don't include this special case. On the other hand, one might argue that the current solution is more robust because it doesn't make assumptions about timeout equality. So maybe we do need this special case, after all.

There is also a related question: do we really need to locate the extremum at every iteration? This looks like an avoidable [Shliemel the painter](https://www.joelonsoftware.com/2001/12/11/back-to-basics/) problem. What we have right now is good enough but perhaps an explanatory comment is warranted, saying that it should be possible to avoid extremum call at every iteration by simply fetching the neighboring node from the current one before deleting it. Obviously this will only work if we rely on the assumption that the timeout is the same for all frames in the transfer (otherwise the loop below will invalidate the tree linkage).

Reply:
While the current solution is acceptable, I must point out that in real-time programming, one should focus not only on the general case but also on the worst case.